### PR TITLE
Add pyzmq

### PR DESCRIPTION
**Pull request by pelson**
I haven't done anything with this other than lift it from conda/conda-recipes. Just wanted to see how it does on windows (predominantly).

### CODE REVIEW

**Comment by pelson** [comment]
@minrk, just wondering if you've seen this before? :

```
===== pyzmq-14.5.0-py27_1 OK =====
+ nosetests zmq.tests
.................SSSSSSSSSSSSSSSS...F................S..................................S.....................................................................S.................
======================================================================
FAIL: test_new (zmq.tests.test_constants.TestConstants)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/conda/envs/_test/lib/python2.7/site-packages/zmq/tests/test_constants.py", line 86, in test_new
    self.fail("AttributeError: zmq.%s" % name)
AssertionError: AttributeError: zmq.IDENTITY_FD

----------------------------------------------------------------------
Ran 176 tests in 24.406s

FAILED (SKIP=19, failures=1)
TESTS FAILED: pyzmq-14.5.0-py27_1
```

**Comment by pelson** [comment]
Without the ``DYLD_LIBRARY_PATH`` I'm able to consistently reproduce the following:

```
Configure: Autodetecting ZMQ settings...
    Custom ZMQ dir:       /Users/pelson/miniconda/envs/_build
gcc -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -arch x86_64 -I/Users/pelson/miniconda/envs/_build/include -Izmq/utils -Izmq/backend/cython -Izmq/devices -c build/temp.macosx-10.5-x86_64-2.7/scratch/vers.c -o build/temp.macosx-10.5-x86_64-2.7/scratch/vers.o
gcc -arch x86_64 -undefined dynamic_lookup build/temp.macosx-10.5-x86_64-2.7/scratch/vers.o -L/Users/pelson/miniconda/envs/_build/lib -lzmq -o build/temp.macosx-10.5-x86_64-2.7/scratch/vers
Error running version detection script:

dyld: Library not loaded: @rpath/libzmq.5.dylib
  Referenced from: /Users/pelson/miniconda/conda-bld/work/pyzmq-14.5.0/build/temp.macosx-10.5-x86_64-2.7/scratch/vers
  Reason: image not found


error: Error running version detection script:

dyld: Library not loaded: @rpath/libzmq.5.dylib
  Referenced from: /Users/pelson/miniconda/conda-bld/work/pyzmq-14.5.0/build/temp.macosx-10.5-x86_64-2.7/scratch/vers
  Reason: image not found


Fatal: Falling back on bundled libzmq, but setup.cfg has explicitly prohibited building the libzmq extension.
```

Answer: set the ``DYLD_LIBRARY_PATH``

**Comment by pelson** [comment]
Seem to be struggling to pick up the zeromq library on Windows:

```
C:\Miniconda\conda-bld\work\pyzmq-14.5.0>"C:\Miniconda\envs\_build\python.exe" setup.py configure --zmq "C:\Miniconda\envs\_build\Library" 
running configure
Warning: Could not copy libzmq into zmq/, which is usually necessary on Windows.Please specify zmq prefix via configure --zmq=/path/to/zmq or copy libzmq into zmq/ manually.
************************************************
Configure: Autodetecting ZMQ settings...
    Custom ZMQ dir:       C:\Miniconda\envs\_build\Library

````

I've double checked the zeromq being shipped with it:

```
win-64-zeromq-4.1.3-vc9_0> find *
Library
Library/AUTHORS.txt
Library/bin
Library/bin/inproc_lat.exe
Library/bin/inproc_thr.exe
Library/bin/libzmq--mt-4_2_0.dll
Library/bin/local_lat.exe
Library/bin/local_thr.exe
Library/bin/remote_lat.exe
Library/bin/remote_thr.exe
Library/include
Library/include/zmq.h
Library/include/zmq_utils.h
Library/lib
Library/lib/libzmq--mt-4_2_0.lib
Library/lib/libzmq-static--mt-s-4_2_0.lib
Library/lib/pkgconfig
Library/lib/pkgconfig/libzmq.pc
```

**Comment by pelson** [comment]
Found the issue. Looks like pyzmq is looking for a ```lib\libzmq.dll``` and we have a ```bin\libzmq--mt-4_2_0.dll``` (https://github.com/zeromq/pyzmq/blob/v14.5.0/setup.py#L322).

**Comment by msarahan** [comment]
Why are there two dashes in the filename? That's fishy, like something isn't being evaluated correctly.

**Comment by pelson** [comment]
> Why are there two dashes in the filename? That's fishy, like something isn't being evaluated correctly.

Indeed. That one is from the anaconda channel (we don't currently build zeromq on Windows).

I've just pushed the change so that pyzmq builds with the bundled version. One day it would be nice to express the dependences properly, but for now, I can live with bundling on Windows.

**Comment by msarahan** [comment]
Yes, that was what I ended up doing also. I tried doing it with separate libraries back in September/October, but eventually gave up.  One day...

**Comment by pelson** [comment]
It gets worse. On py35 I get the following error when bundling on Windows:

```
LINK : fatal error LNK1181: cannot open input file 'libzmq.lib'
error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio 14.0\\VC\\BIN\\link.exe' failed with exit status 1181
 
C:\Miniconda\conda-bld\work\pyzmq-14.5.0>if errorlevel 1 exit 1 
Command failed: C:\Windows\system32\cmd.exe /c call bld.bat
```

Ironically, this one worked fine without bundling... I'm going to put a conditional in the build so that *some* windows builds bundle and others link.

**Comment by msarahan** [comment]
I don't remember that problem.  Let me see about giving you our current recipe before you go down that road.

File: recipes/pyzmq/bld.bat [hunk 1]
```diff
@@ -0,0 +1,13 @@
+set DISTUTILS_USE_SDK=1
+
+"%PYTHON%" setup.py configure --zmq=bundled
+if errorlevel 1 exit 1
+
+
+# For now, we build pyzmq with the bundled zeromq. There were a number of reasons for
+# this, including lib\libzmq.dll not being available, as well as
```
Comment (msarahan) [inline recipes/pyzmq/bld.bat]: # is an invalid comment character in bat files.  you need

REM comment
or
:: comment

**Comment by jakirkham** [comment]
Yeah, I'm sure this is getting sorted, but @minrk designed `pyzmq` on Windows to be self-contained with a vanilla install. It does not use an external `zeromq` by default. Also, it does not use `libsodium`, but comes with tweetnacl.

**Comment by jakirkham** [comment]
If you would like, I can play with this on my Mac and submit a PR to your branch, @pelson. Just let me know.

**Comment by pelson** [comment]
> If you would like, I can play with this on my Mac and submit a PR to your branch, @pelson. Just let me know.

Sure thing. I was working this morning - just need to remember how I had it set up! :smile: 
I hacked around and added gcc - that certainly wasn't there this morning though.

**Comment by jakirkham** [comment]
Cool, I'll get right on it then.

**Comment by jakirkham** [comment]
Thanks again for diving into getting all of this Jupyter stuff packaged. This is going to be very helpful. I am also interested in this so if you want to divvy up the labor somehow please let me know your thoughts.

**Comment by minrk** [comment]
The pyzmq bundling on Windows is really a punt because I couldn't figure out reliable linking on Windows (that code was mostly settled before conda was a thing). I would start the pyzmq package only bundling zeromq on Windows, and investigate the external linking as a later improvement. It may require PRs to pyzmq, due to the funky library naming, which I would be happy to accept.

**Comment by jakirkham** [comment]
Definitely, I completely agree with that proposal as a first pass. Ideally, we will want to break them out, but doing that in later iterations seem more appropriate.

**Comment by jakirkham** [comment]
Alright, I have some fixes incoming for you, @pelson. :smile:

**Comment by jakirkham** [comment]
Did you manage to get zeromq to work with Python 3.5 on Windows?

**Comment by jakirkham** [comment]
See this PR ( https://github.com/pelson/staged-recipes/pull/2 ) for Mac fixes. It also uses `clang` now.

File: recipes/pyzmq/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,43 @@
+{% set version = "14.5.0" %}
```
Comment (jakirkham) [inline recipes/pyzmq/meta.yaml]: Currently 15.2.0 is out. Should we upgrade?

Comment (jakirkham) [inline recipes/pyzmq/meta.yaml]: Running into this issue ( https://github.com/conda/conda-build/issues/686 ).

```
Writing /zopt/conda2/envs/_build/lib/python2.7/site-packages/pyzmq-15.2.0-py2.7.egg-info
compiling .pyc files...
Compiling /zopt/conda2/envs/_build/lib/python2.7/site-packages/zmq/asyncio.py ...
  File "/zopt/conda2/envs/_build/lib/python2.7/site-packages/zmq/asyncio.py", line 79
    "{!r}".format(fileobj)) from None
                               ^
SyntaxError: invalid syntax

Compiling /zopt/conda2/envs/_build/lib/python2.7/site-packages/zmq/auth/asyncio.py ...
  File "/zopt/conda2/envs/_build/lib/python2.7/site-packages/zmq/auth/asyncio.py", line 28
    events = yield from self.__poller.poll()
                      ^
SyntaxError: invalid syntax

Compiling /zopt/conda2/envs/_build/lib/python2.7/site-packages/zmq/tests/_test_asyncio.py ...
  File "/zopt/conda2/envs/_build/lib/python2.7/site-packages/zmq/tests/_test_asyncio.py", line 43
    yield from a.send(b'hi')
             ^
SyntaxError: invalid syntax

Command failed: /zopt/conda2/envs/_build/bin/python -Wi /zopt/conda2/envs/_build/lib/python2.7/compileall.py -q -x port_v3 /zopt/conda2/envs/_build/lib/python2.7/site-packages
```

Comment (jakirkham) [inline recipes/pyzmq/meta.yaml]: Was able to update to `14.7.0` though. That also fixed the test failure. :tada:

**Comment by jakirkham** [comment]
Updated to the latest 14 release (14.7.0) (was having trouble building the latest 15 release due to this bug ( https://github.com/conda/conda-build/issues/686 )). This fixed the failing test so the test suite is enabled. This is all in my PR.

**Comment by pelson** [comment]
> Updated to the latest 14 release (14.7.0) (was having trouble building the latest 15 release due to this bug ( conda/conda-build#686 )). This fixed the failing test so the test suite is enabled. This is all in my PR.

Amazing. Thanks @jakirkham! Upgrading to 15 should be part of the feedstock I feel. @msarahan provided Continuum's latest recipe in https://github.com/pelson/staged-recipes/pull/1 which needed to deal with the py3 syntax also. I didn't want to cross that bridge in the staged-recipes PR, and think the feedstock it is the right place to be having the discussion.

**Comment by jakirkham** [comment]
That unable to find `clang` error is pretty weird. I probably would have just tried to restart it.

File: recipes/pyzmq/build.sh [hunk 3]
```diff
@@ -1,9 +1,10 @@
 #!/bin/bash
 
-export DYLD_LIBRARY_PATH="$PREFIX/lib:$DYLD_LIBRARY_PATH"
```
Comment (jakirkham) [inline recipes/pyzmq/build.sh]: We really need this line for Mac unfortunately. I was unable to build locally without it either.

**Comment by jakirkham** [comment]
Yeah, I think we need `export DYLD_LIBRARY_PATH="$PREFIX/lib:$DYLD_LIBRARY_PATH"` back. I was unable to build locally without this.

**Comment by ocefpaf** [comment]
@pelson I cancelled Appveyor because it is hanging waiting for an answer:

```shell
===== pyzmq-14.7.0-py27_0 OK =====
 
C:\Miniconda\conda-bld\test-tmp_dir>nosetests zmq.tests 
.................S.................................S.......S......................................SSSSSS...................SS.....S........... 
----------------------------------------------------------------------
Ran 180 tests in 29.081s
 
OK (SKIP=28)
Terminate batch job (Y/N)? 
```

**Comment by pelson** [comment]
The only problem remains with OSX based on the following:

```
running configure
clang -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -arch x86_64 -I/Users/travis/miniconda/envs/_build/include -c build/temp.macosx-10.5-x86_64-2.7/scratch/check_sys_un.c -o build/temp.macosx-10.5-x86_64-2.7/scratch/check_sys_un.o
sh: line 1:   740 Trace/BPT trap: 5       /Applications/Xcode.app/Contents/Developer/usr/bin/xcodebuild -sdk / -find clang 2> /dev/null
clang: error: unable to find utility "clang", not a developer tool or in PATH
Warning: No sys/un.h, IPC_PATH_MAX_LEN will be undefined: command 'clang' failed with exit status 72
************************************************
Configure: Autodetecting ZMQ settings...
    Custom ZMQ dir:       /Users/travis/miniconda/envs/_build
clang -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -arch x86_64 -I/Users/travis/miniconda/envs/_build/include -Izmq/utils -Izmq/backend/cython -Izmq/devices -c build/temp.macosx-10.5-x86_64-2.7/scratch/vers.c -o build/temp.macosx-10.5-x86_64-2.7/scratch/vers.o
sh: line 1:   746 Trace/BPT trap: 5       /Applications/Xcode.app/Contents/Developer/usr/bin/xcodebuild -sdk / -find clang 2> /dev/null
clang: error: unable to find utility "clang", not a developer tool or in PATH
error: command 'clang' failed with exit status 72
Fatal: Falling back on bundled libzmq, but setup.cfg has explicitly prohibited building the libzmq extension.
Command failed: /bin/bash -x -e /Users/travis/build/conda-forge/staged-recipes/recipes/pyzmq/build.sh
```

This is despite the ability to which clang:

```
+ which clang
/usr/bin/clang
```

I'm confident that this originally built, but can't remember exactly what has changed (I'm considering whether I've rebased and am now including the removal of all of homebrew's packages).

**Comment by pelson** [comment]
Thanks @ocefpaf - I'd forgotten about that one. Phew, this one is testing my patience 😄  - I'll be glad to get it done!

**Comment by pelson** [comment]
Windows and Linux now passing. OSX is proving to be more problematic. I have reason to believe that by removing ```xctools``` from homebrew we are forcing the use of the TravisCI provided Apple standard ```xcodebuild```, which seems to be broken. I'm going to try not removing the ```xctools``` brew package.

**Comment by pelson** [comment]
Seems the xcodebuild is broken on TravisCI:

```
+ xcodebuild -sdk / -find clang
dyld: Symbol not found: _sqlite3_intarray_bind
  Referenced from: /System/Library/Frameworks/CoreData.framework/Versions/A/CoreData
  Expected in: /Users/travis/miniconda/envs/_build/lib/libsqlite3.dylib
 in /System/Library/Frameworks/CoreData.framework/Versions/A/CoreData
/Users/travis/build/conda-forge/staged-recipes/recipes/pyzmq/build.sh: line 20:   740 Trace/BPT trap: 5       xcodebuild -sdk / -find clang
```

I'm going to try to avoid removing some of the brew dependencies to see if that fixes it.

**Comment by minrk** [comment]
The Trace/BPT trap may be related to setting DYLD_LIBRARY_PATH, I frequently see those errors when DYLD path is set. If it's needed for some part (either test or libzmq detection), you might try setting it only for those commands, rather than from the beginning.

**Comment by pelson** [comment]
The weird thing is that I can't reproduce the problem on my own machine. I suspect there is an interplay between the homebrew removal and the DYLD library path which is worth understanding (although this is turning into a bit of a odyssey 😄 ).

File: recipes/pyzmq/build.sh [hunk 4]
```diff
@@ -0,0 +1,25 @@
+#!/bin/bash
+
+ZMQ="${PREFIX}"
+
+if [ "$(uname)" == "Darwin" ]; then
+    ZMQ=bundled
+    export MACOSX_VERSION_MIN=10.7
+    export CC="clang"
+    export MACOSX_DEPLOYMENT_TARGET=10.6
+    export DYLD_LIBRARY_PATH="$PREFIX/lib:$DYLD_LIBRARY_PATH"
```
Comment (minrk) [inline recipes/pyzmq/build.sh]: moving this line to the end should prevent it from interfering with the `xcodebuild` command

File: recipes/pyzmq/build.sh [hunk 5]
```diff
@@ -0,0 +1,25 @@
+#!/bin/bash
+
```
Comment (minrk) [inline recipes/pyzmq/build.sh]: `set -e` here for early exit on error?

Comment (minrk) [inline recipes/pyzmq/build.sh]: ...or is that already conda-build's default behavior?

**Comment by pelson** [comment]
Just wanted to state that homebrew was a red-herring. The build still fails (same libsqlite3 issue) even when we don't remove all homebrew dependencies.

Indeed, on my own machine:

```
DYLD_LIBRARY_PATH=~/miniconda/envs/_test/lib xcodebuild -sdk / -find clang
dyld: Symbol not found: _sqlite3_intarray_bind
  Referenced from: /System/Library/Frameworks/CoreData.framework/Versions/A/CoreData
  Expected in: /Users/pelson/miniconda/envs/_test/lib/libsqlite3.dylib
 in /System/Library/Frameworks/CoreData.framework/Versions/A/CoreData
Trace/BPT trap: 5
```

Phew. I'll follow @minrk's advice, and try to figure out where DYLD_LIBRARY_PATH is actually needed.

File: recipes/pyzmq/build.sh (hunk 5 above)
Comment (pelson) [inline recipes/pyzmq/build.sh]: Already conda-build's behaviour on Linux/OSX. On Windows it (the equivalent behaviour) isn't the case though.

**Comment by pelson** [comment]
Ok, looks like I've managed to workaround the need for the DYLD_LIBRARY_PATH, and all platforms are now building as expected. 🎉 

I've added @minrk, @msarahan and @jakirkham as maintainers to this recipe. The more the merrier IMO. There is a lot of history in this issue which I will link against in the [pyzmq feedstock](https://github.com/conda-forge/pyzqm-feedstock).

**Pull request merged**